File: java/parser/src/main/java/gov/lbl/parser/parser/ExpressionVisitor.java
package gov.lbl.parser.parser;

import java.util.List;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Expression;
import gov.lbl.parser.domain.If_elseif_expression;
import gov.lbl.parser.domain.If_expression;
import gov.lbl.parser.domain.Simple_expression;

import static java.util.stream.Collectors.toList;

import java.util.ArrayList;

public class ExpressionVisitor extends modelicaBaseVisitor<Expression> {
    @Override
    public Expression visitExpression(modelicaParser.ExpressionContext ctx) {
        Simple_expressionVisitor simple_expressionVisitor = new Simple_expressionVisitor();
        Simple_expression simple_expression = ctx.simple_expression() == null ? null : ctx.simple_expression().accept(simple_expressionVisitor);

        ExpressionVisitor expressionVisitor = new ExpressionVisitor();
        List<Expression> expressions = ctx.expression() == null ? null : ctx.expression()
                                                                            .stream()
                                                                            .map(expr -> expr.accept(expressionVisitor))
                                                                            .collect(toList());
        List<If_elseif_expression> if_elseif = new ArrayList<If_elseif_expression>();
        for (int i = 0; i<expressions.size() - 1; i+=2) {
            Expression condition = expressions.get(i);
            Expression then = expressions.get(i+1);
            if_elseif.add(new If_elseif_expression(condition, then));
        }
        // expressions always have else? 
        Expression else_expression = expressions.get(expressions.size()-1);

        If_expression if_expression = new If_expression(if_elseif, else_expression);
        return new Expression(simple_expression, if_expression);
    }
}

File: jsParser/parser/ExpressionVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Expression = require('../domain/Expression')
const If_expression = require('../domain/If_expression')
const If_elseif_expression = require('../domain/If_elseif_expression')

const Simple_expressionVisitor = require('./Simple_expressionVisitor')

class ExpressionVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitExpression (ctx) {
    let simple_expression = null
    let if_expression = null
    const expressions = []
    const if_elseif = []

    if (ctx.simple_expression()) {
      const simple_expressionVisitor = new Simple_expressionVisitor.Simple_expressionVisitor()
      simple_expression = simple_expressionVisitor.visitSimple_expression(ctx.simple_expression())
    }

    if (ctx.expression()) {
      ctx.expression().forEach(expr => {
        expressions.push(this.visitExpression(expr))
      })
    }

    if (expressions.length > 0) {
      for (let i = 0; i < expressions.length - 1; i += 2) {
        const condition = expressions[i]
        const then = expressions[i + 1]
        if_elseif.push(new If_elseif_expression.If_elseif_expression(condition, then))
      }
      if (expressions.length > 2) {
        const else_expression = expressions[expressions.length - 1]
        if_expression = new If_expression.If_expression(if_elseif, else_expression)
      } else {
        if_expression = new If_expression.If_expression(if_elseif, null)
      }
    }

    return new Expression.Expression(simple_expression, if_expression)
  }
};

ExpressionVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitExpression = this.visitExpression
exports.ExpressionVisitor = ExpressionVisitor
